[JSC] Add ProxyObjectHas IC to optimize "has" trap

shvaikalesh opened the pull request:
#### 48a8c72db72281068a18847926c158d0b49673e1
<pre>
[JSC] Add ProxyObjectHas IC to optimize &quot;has&quot; trap
<a href="https://bugs.webkit.org/show_bug.cgi?id=253830">https://bugs.webkit.org/show_bug.cgi?id=253830</a>
&lt;rdar://problem/106644477&gt;

Reviewed by Yusuke Suzuki.

This change adds ProxyObjectHas IC for Proxy &quot;has&quot; trap, which detects ProxyObject and calls
@performProxyObjectHas JS function that, in the most common case, avoids calling into C++
to validate the trap result.

While progressing provided microbenchmarks by up to 10x, unfortunately this patch does not
affect Speedometer2 nor JetStream3: JS3/proxy-mobx invokes &quot;has&quot; trap only on indexed properties,
which are currently rejected by isCacheableIdentifierCell() and not supported by ProxyObjectAccessCase.

Parameter order of &quot;get&quot; and &quot;set&quot; IC helpers had to be rearranged to simplify ProxyObjectAccessCase::emit().

                                   ToT                       patch

proxy-has-hit                48.4694+-0.3179     ^      4.8357+-0.0314        ^ definitely 10.0232x faster
proxy-has-miss               53.6834+-0.3113     ^     15.9198+-0.0730        ^ definitely 3.3721x faster
proxy-has-miss-handler       30.9117+-0.1541     ^      4.9238+-0.0337        ^ definitely 6.2781x faster

&lt;geometric&gt;                  43.1641+-0.1435     ^      7.2367+-0.0233        ^ definitely 5.9646x faster

* JSTests/microbenchmarks/proxy-has-hit.js: Added.
* JSTests/microbenchmarks/proxy-has-miss-handler.js: Added.
* JSTests/microbenchmarks/proxy-has-miss.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ProxyHelpers.js:
(linkTimeConstant.performProxyObjectHas):
(linkTimeConstant.performProxyObjectGet):
(linkTimeConstant.performProxyObjectSetSloppy):
(linkTimeConstant.performProxyObjectSetStrict):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.cpp:
(JSC::ProxyObjectAccessCase::emit):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheInBy):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handlePutById):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::performProxyObjectHasFunction const):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::performHasProperty):
(JSC::ProxyObject::validateNegativeHasTrapResult):
* Source/JavaScriptCore/runtime/ProxyObject.h:

Canonical link: <a href="https://commits.webkit.org/261628@main">https://commits.webkit.org/261628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19194d165eb87f088c8e2cfc346288f64c572437

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112276 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/907 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4066 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120903 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12565 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4839 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118039 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105333 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/632 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45921 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100665 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13801 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/677 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/94276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11933 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14489 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102030 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52681 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31895 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16286 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110074 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4417 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27196 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/AccessCase.cpp:
```diff
@@ -82,6 +82,7 @@ Ref<AccessCase> AccessCase::create(VM& vm, JSCell* owner, AccessType type, Cache
     case DirectArgumentsLength:
     case ScopedArgumentsLength:
     case ModuleNamespaceLoad:
+    case InProxyObject:
```
We should name it `ProxyObjectHas` / `ProxyObjectIn` to name all proxy-related-IC starting with `ProxyObject`.

Constellation commented on Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.cpp:
```diff
@@ -115,29 +130,27 @@ void ProxyObjectAccessCase::emit(AccessGenerationState& state, MacroAssembler::J
+        jit.storeCell(thisGPR, calleeFrame.withOffset(virtualRegisterForArgumentIncludingThis(2).offset() * sizeof(Register)));
 
     if (m_type == ProxyObjectStore)
         jit.storeValue(valueRegs, calleeFrame.withOffset(virtualRegisterForArgumentIncludingThis(3).offset() * sizeof(Register)));
```
I think this is hard to follow. We should just set up arguments in the switch for each case. This many if branchings make the code more unreadable.

Constellation commented on Source/JavaScriptCore/runtime/ProxyObject.cpp:
```diff
@@ -341,25 +341,33 @@ bool ProxyObject::performHasProperty(JSGlobalObject* globalObject, PropertyName
-            }
+    if (trapResultAsBool)
+        return true;
+
```
scope.release() is required.

shvaikalesh replied:
Oh nice catch, thanks, fixed.

Constellation reviewed (approved):
r=me

Commit:
[JSC] Add ProxyObjectHas IC to optimize "has" trap
https://bugs.webkit.org/show_bug.cgi?id=253830
<rdar://problem/106644477>

Reviewed by Yusuke Suzuki.

This change adds ProxyObjectHas IC for Proxy "has" trap, which detects ProxyObject and calls
@performProxyObjectHas JS function that, in the most common case, avoids calling into C++
to validate the trap result.

While progressing provided microbenchmarks by up to 10x, unfortunately this patch does not
affect Speedometer2 nor JetStream3: JS3/proxy-mobx invokes "has" trap only on indexed properties,
which are currently rejected by isCacheableIdentifierCell() and not supported by ProxyObjectAccessCase.

Parameter order of "get" and "set" IC helpers had to be rearranged to simplify ProxyObjectAccessCase::emit().

                                   ToT                       patch

proxy-has-hit                48.4694+-0.3179     ^      4.8357+-0.0314        ^ definitely 10.0232x faster
proxy-has-miss               53.6834+-0.3113     ^     15.9198+-0.0730        ^ definitely 3.3721x faster
proxy-has-miss-handler       30.9117+-0.1541     ^      4.9238+-0.0337        ^ definitely 6.2781x faster

<geometric>                  43.1641+-0.1435     ^      7.2367+-0.0233        ^ definitely 5.9646x faster

* JSTests/microbenchmarks/proxy-has-hit.js: Added.
* JSTests/microbenchmarks/proxy-has-miss-handler.js: Added.
* JSTests/microbenchmarks/proxy-has-miss.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ProxyHelpers.js:
(linkTimeConstant.performProxyObjectHas):
(linkTimeConstant.performProxyObjectGet):
(linkTimeConstant.performProxyObjectSetSloppy):
(linkTimeConstant.performProxyObjectSetStrict):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.cpp:
(JSC::ProxyObjectAccessCase::emit):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheInBy):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handlePutById):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::performProxyObjectHasFunction const):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::performHasProperty):
(JSC::ProxyObject::validateNegativeHasTrapResult):
* Source/JavaScriptCore/runtime/ProxyObject.h:

Canonical link: https://commits.webkit.org/261628@main
